ts/Extensions/DataGrouping.ts

```diff
@@ -508,6 +528,126 @@ const groupData = function (
+        totalRange = series.currentDataGrouping && series.currentDataGrouping.gapSize;
+    let i;
+
+    // DataGrouping x-coordinates.
+    if (dataGroupingOptions && series.xData && totalRange && series.groupMap) {
+        const groupedDataLength = groupedXData.length - 1,
+            anchor = dataGroupingOptions.anchor,
+            firstAnchor = pick(dataGroupingOptions.firstAnchor, anchor),
+            lastAnchor = pick(dataGroupingOptions.lastAnchor, anchor);
+
+        // Anchor points that are not extremes.
+        if (anchor && anchor !== 'start') {
+            const shiftInterval: number = totalRange *
+                ({ middle: 0.5, end: 1 } as Highcharts.anchorChoiceType)[anchor];
+
+            i = groupedXData.length - 1;
+            while (i-- && i > 0) {
+                groupedXData[i] += shiftInterval;
+            }
+        }
+
+        // Change the first point position, but only when it is
+        // the first point in the data set not in the current zoom.
+        if (
+            firstAnchor &&
+            firstAnchor !== 'start' &&
+            series.xData[0] >= groupedXData[0]
+        ) {
+            const groupStart = series.groupMap[0].start,
+                groupLength = series.groupMap[0].length;
+            let firstGroupstEnd;
+
+            if (isNumber(groupStart) && isNumber(groupLength)) {
+                firstGroupstEnd = groupStart + (groupLength - 1);
+            }
+
+            groupedXData[0] = ({
+                middle: groupedXData[0] + 0.5 * totalRange,
+                end: groupedXData[0] + totalRange,
+                firstPoint: series.xData[0],
+                lastPoint: firstGroupstEnd && series.xData[firstGroupstEnd]
+            } as Highcharts.anchorChoiceType)[firstAnchor];
+        }
+
+        // Change the last point position but only when it is
+        // the last point in the data set not in the current zoom.
+        if (
+            lastAnchor &&
+            lastAnchor !== 'start' &&
+            totalRange &&
+            groupedXData[groupedDataLength] >= xMax - totalRange
+        ) {
+            const lastGroupStart = series.groupMap[series.groupMap.length - 1].start;
+
+            groupedXData[groupedDataLength] = ({
+                middle: groupedXData[groupedDataLength] + 0.5 * totalRange,
+                end: groupedXData[groupedDataLength] + totalRange,
+                firstPoint: lastGroupStart && series.xData[lastGroupStart],
+                lastPoint: series.xData[series.xData.length - 1]
+            } as Highcharts.anchorChoiceType)[lastAnchor];
+        }
+    }
+};
+
+const adjustExtremes = function (
+    xAxis: Axis,
+    series: Series,
+    groupedXData: Array<number>
+): void {
+    // Make sure the X axis extends to show the first group (#2533)
+    // But only for visible series (#5493, #6393)
+    if (
+        defined(groupedXData[0]) &&
+        groupedXData[0] < (xAxis.min as any)
```

Comment:
Avoid any-casting of `xAxis.min` and `xAxis.dataMin`. Add a typecheck, `isNumber(axis.min)` etc., to avoid unpleasant surprises when `min` is undefined.